File: Programs/RCTNative/RCTNative/TabBar.m
//
//  TabBar.m
//  RCTNative
//
//  Created by 杨涵 on 16/6/24.
//  Copyright © 2016年 yanghan. All rights reserved.
//

#import "TabBar.h"

@interface TabBar()

@property (nonatomic, strong)   NSArray         *btnImages;
@property (nonatomic, strong)   NSMutableArray         *buttons;
@property (nonatomic, strong)   UIButton        *selButton;

@end

@implementation TabBar

- (NSArray *)btnImages {
    
    return @[@"tabbar_home.png",@"tabbar_message.png",@"tabbar_profile.png"];
}

- (NSMutableArray *)buttons {
    if (_buttons == nil) {
        _buttons = [NSMutableArray array];
    }
    return _buttons;
}

- (void)awakeFromNib {
    [super awakeFromNib];
    
    for (NSInteger i=0; i<self.btnImages.count; i++) {
        NSString *imageName = self.btnImages[i];
        UIButton *button = [self buttonWithImageName:imageName];
        button.tag = i;
        if (i == 0) {
            [self buttonClick:button];
        }
        [self addSubview:button];
        [self.buttons addObject:button];
    }
}

- (void)layoutSubviews {
    [super layoutSubviews];
    
    NSInteger count = self.buttons.count;
    double width = CGRectGetWidth(self.bounds) / count;
    double height = CGRectGetHeight(self.bounds);
    
    CGRect frame = CGRectMake(0, 0, width, height);
    for (UIButton *btn in self.buttons) {
        btn.frame = CGRectOffset(frame, btn.tag*width, 0);
    }
}

- (UIButton *)buttonWithImageName:(NSString *)imageName {
    UIButton *button = [UIButton buttonWithType:UIButtonTypeCustom];
    [button setImage:[UIImage imageNamed:imageName] forState:UIControlStateNormal];
    [button addTarget:self action:@selector(buttonClick:) forControlEvents:UIControlEventTouchUpInside];
    return button;
}

- (void)buttonClick:(UIButton *)sender {
    if (sender == self.selButton) {
        return;
    }
    
    self.selButton = sender;
    [self.adelegate tabBar:self didClickButton:sender.tag];
}
@end

File: Programs/Connection/Connection/TabBar.swift
//
//  TabBar.swift
//  Connection
//
//  Created by 杨涵 on 16/4/11.
//  Copyright © 2016年 yanghan. All rights reserved.
//

import UIKit

protocol TabBarDelegate: NSObjectProtocol {
    /**
     点击某个按钮
     -parameter tabBar:tabBar
     -parameter index:按钮index
    */
    func tabBar(tabBar: TabBar, didClickButton index: Int)
}

class TabBar: UITabBar {
    weak var aDelegate: TabBarDelegate?
    
    private var selButton: DHButton?
    
    private lazy var btnImages: [String] = {
        return ["tabbar_home", "tabbar_discover", "tabbar_show", "tabbar_message", "tabbar_profile"]
    }()
    
    private lazy var buttons: [UIButton] = {
        return [UIButton]()
    }()
    
    override func awakeFromNib() {
        super.awakeFromNib()
        
        for i in 0..<btnImages.count {
            let imageName = btnImages[i]
            let b = buttonWithImageName(imageName)
            b.tag = i
            if i == 0 {
                buttonClick(b)
            }
            addSubview(b)
            buttons.append(b)
        }
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        
        let count = buttons.count
        let w = CGRectGetWidth(bounds) / CGFloat(count)
        let h = CGRectGetHeight(bounds)
        
        let frame = CGRectMake(0, 0, w, h)
        for b in buttons {
            b.frame = CGRectOffset(frame, CGFloat(b.tag)*w, 0)
        }
    }
    
    private func buttonWithImageName(imageName: String) -> DHButton {
        let b = DHButton(type: .Custom)
        let sImageName = imageName + "_sel"
        b.setImage(UIImage(named: imageName), forState: .Normal)
        b.setImage(UIImage(named: sImageName), forState: .Selected)
        b.addTarget(self, action: #selector(TabBar.buttonClick(_:)), forControlEvents: .TouchDown)
        return b
    }
    
    func buttonClick(button: DHButton) {
        guard selButton != button else {
            return
        }
        
        if button.tag != 2 {
            selButton?.selected = false
            button.selected = true
            selButton = button
        }
        
        aDelegate?.tabBar(self, didClickButton: button.tag)
    }
}
